DATA-35 - Fixing support for MySQL and MongoDB data.

Squirrel18 opened the pull request:
## Description:

This PR fixes the support for the MySQL and MongoDB data generated from edx-analytics-exporter.
The opaque-keys package was added in order to convert new and deprecated course ids.

sqlfromS3 command was added to obatin the edx-analytics-exporter data from Amazon S3, other commands were modified to support local files instead of obtaining them from Google Cloud Storage.

## Reviewers:

- [ ] @andrey-canon 
- [ ] @diegomillan

andrey-canon commented on edx2bigquery/do_waldofication_of_sql.py:
```diff
@@ -101,16 +108,18 @@ def fix_course_id(short_cid, date):
             os.mkdir(cdir)
     else:
-        if not cid in known_course_ids or []:
+        if not course_id in known_course_ids or []:
```
`or []` is unnecessary

Squirrel18 replied:
It's part of the exisiting code, but, it's addressed.

andrey-canon commented on edx2bigquery/edx2course_axis.py:
```diff
@@ -627,7 +627,16 @@ def process_course(dir, use_dataset_latest=False, force_course_id=None):
+                cdat,
+                caset,
+                use_local_files,
+                ret[default_cid]['bundle'],
```
This raises an exception, `'bundle'` should be a constant

Squirrel18 replied:
Addressed.

andrey-canon commented on edx2bigquery/s3_backend.py:
```diff
@@ -22,32 +23,31 @@ def get_simple_storage_service_client():
 
-    if not os.path.exists(logs_dir):
-        os.mkdir(logs_dir)
+    bucket_name = getattr(edx2bigquery_config, 'AWS_BUCKET_NAME', '')
```
`'AWS_BUCKET_NAME'` should be a constant

Squirrel18 replied:
This is a setting name, so, it's not neccesary.

andrey-canon commented on edx2bigquery/main.py:
```diff
@@ -481,6 +491,25 @@ def get_logs_from_simple_storage_service(param, args):
+        raise Exception('--start-date must be specified in this format: YYYY-MM-DD.')
+
+    get_sql_data_objects(
+        bucket_name=getattr(edx2bigquery_config, 'AWS_BUCKET_NAME', None),
```
`'AWS_BUCKET_NAME'` should be a constant

Squirrel18 replied:
This is a setting name, so, it's not neccesary.

andrey-canon commented on edx2bigquery/bqutil.py:
```diff
@@ -788,13 +787,14 @@ def upload_local_data_to_big_query(dataset_id, table_id, schema, course_id, file
         schema: Schema type to use upon the data.
         course_id: Valid course id of the current proccessed course.
         file_name: File name of the archive that has the tracking log data.
+        source_format: Source format of the file data e.g. JSON, CSV
```
Period at the end of the line

Squirrel18 replied:
Addressed.

andrey-canon commented on edx2bigquery/bqutil.py:
```diff
@@ -809,29 +809,45 @@ def upload_local_data_to_big_query(dataset_id, table_id, schema, course_id, file
 
     Args:
         schema: Schema type to use upon the data.
+        source_format: Source format of the file data e.g. JSON, CSV
```
period at the end

Squirrel18 replied:
Addressed.

andrey-canon commented on edx2bigquery/bqutil.py:
```diff
@@ -809,29 +809,45 @@ def upload_local_data_to_big_query(dataset_id, table_id, schema, course_id, file
+        job_config.source_format = bigquery.SourceFormat.NEWLINE_DELIMITED_JSON
+    elif source_format == 'CSV':
+        job_config.source_format = bigquery.SourceFormat.CSV
+        job_config.skip_leading_rows = 1
```
what happen if source_format is different to CSV and SON

Squirrel18 replied:
https://cloud.google.com/bigquery/docs/loading-data#supported_data_formats The default source format for loading data is CSV, if the source format is different than CSV, the Google BigQuery job will fail.

andrey-canon commented on edx2bigquery/load_course_sql.py:
```diff
@@ -389,4 +397,58 @@ def copy_if_newer(fn, fnset, options='-z csv,json'):
+    for table_name, scheme_data in sql_scheme_file_names.items():
+        dict_schema = get_schema_from_file(scheme_data.get('scheme_file_name', ''))
+
+        if table_name == 'studentmodule':
```
this should be a constant

Squirrel18 replied:
Addressed.

andrey-canon commented on edx2bigquery/load_course_sql.py:
```diff
@@ -389,4 +397,58 @@ def copy_if_newer(fn, fnset, options='-z csv,json'):
+
+        if table_name == 'studentmodule':
+            scheme = dict_schema.get('courseware_studentmodule', None)
+        elif table_name == 'user_info_combo':
```
this should be a constant

Squirrel18 replied:
Addressed.

andrey-canon commented on edx2bigquery/load_course_sql.py:
```diff
@@ -389,4 +397,58 @@ def copy_if_newer(fn, fnset, options='-z csv,json'):
+        dict_schema = get_schema_from_file(scheme_data.get('scheme_file_name', ''))
+
+        if table_name == 'studentmodule':
+            scheme = dict_schema.get('courseware_studentmodule', None)
```
None is the default value

Squirrel18 replied:
Discussion addressed. Explicit is better than implicit.

andrey-canon replied:
The None in .get(..., None) is really redundant and superfluous. taken from https://stackoverflow.com/questions/46546872/what-is-the-scenario-of-dict-get-none-or-none
It is a little superfluous in this case, as get() defaults to None for the second parameter. taken from https://stackoverflow.com/questions/12631620/what-does-this-none-do-in-get-in-python

Squirrel18 replied:
Sorry, my dafault value should not be None, it should be a dict, since shema is a dict. Addressed.

andrey-canon commented on edx2bigquery/load_course_sql.py:
```diff
@@ -389,4 +397,58 @@ def copy_if_newer(fn, fnset, options='-z csv,json'):
+        if table_name == 'studentmodule':
+            scheme = dict_schema.get('courseware_studentmodule', None)
+        elif table_name == 'user_info_combo':
+            scheme = dict_schema.get('user_info_combo', None)
```
None is the default value

Squirrel18 replied:
Discussion addressed. Explicit is better than implicit.

andrey-canon replied:
The None in .get(..., None) is really redundant and superfluous. taken from https://stackoverflow.com/questions/46546872/what-is-the-scenario-of-dict-get-none-or-none
It is a little superfluous in this case, as get() defaults to None for the second parameter. taken from https://stackoverflow.com/questions/12631620/what-does-this-none-do-in-get-in-python

Squirrel18 replied:
Sorry, my dafault value should not be None, it should be a dict, since shema is a dict. Addressed.

andrey-canon commented on edx2bigquery/load_course_sql.py:
```diff
@@ -389,4 +397,58 @@ def copy_if_newer(fn, fnset, options='-z csv,json'):
+            schema=scheme,
+            course_id=course_id,
+            file_name=file_name,
+            source_format='CSV',
```
this should be a constant

Squirrel18 replied:
Addressed.

andrey-canon commented on edx2bigquery/load_daily_tracking_logs.py:
```diff
@@ -163,6 +163,7 @@ def load_local_logs_to_biqquery(course_id, start_date, end_date, verbose):
                 schema=schema,
                 course_id=course_id,
                 file_name=file_name,
+                source_format='JSON',
```
this should be a constant

Squirrel18 replied:
Addressed.

andrey-canon commented on edx2bigquery/main.py:
```diff
@@ -2202,6 +2253,9 @@ def CommandLine():
     elif args.command == 'logsfromS3':
         get_logs_from_simple_storage_service(param, args)
 
+    elif args.command == 'sqlfromS3':
```
this should be a constant

Squirrel18 replied:
No, since these are the names of the command, and we can't implement it with the existing code.

andrey-canon commented on edx2bigquery/make_course_axis.py:
```diff
@@ -18,11 +21,7 @@ def process_course(course_id, basedir, datedir, use_dataset_latest, verbose=Fals
-                'course-prod-edge-analytics.xml.tar.gz',
-                'course-prod-edx-replica.xml.tar.gz',
-            ]
+    fn_to_try = getattr(edx2bigquery_config, 'COURSE_FILES_PREFIX_NAMES', [])
```
this should be a constant

Squirrel18 replied:
This is a setting name, so, it's not neccesary.

andrey-canon commented on edx2bigquery/make_person_course.py:
```diff
@@ -55,22 +55,25 @@
 
-csv.field_size_limit(13107200)
 
+csv.field_size_limit(13107200)
```
this should be a constant

Squirrel18 replied:
This comes from the original implementation, so, ignore it.

andrey-canon commented on edx2bigquery/make_video_analysis.py:
```diff
@@ -126,11 +130,22 @@ def make_video_stats(course_id, api_key, basedir, datedir, force_recompute, use_
+                the_schema,
+                course_id,
+                fileoutput,
+                'JSON',
```
this should be a constant

Squirrel18 replied:
Addressed.

andrey-canon commented on edx2bigquery/s3_backend.py:
```diff
@@ -109,4 +113,67 @@ def get_tracking_log_objects(bucket_name, start_date):
+
+    # .zip Since edx-analytics-exporter uploads a .zip file containing all the generated data.
+    object_key_name = '{}/{}{}{}'.format(
+        getattr(edx2bigquery_config, 'SQL_DATA_BUCKET_PATH', ''),
```
constant

Squirrel18 replied:
This is a setting name, so, it's not neccesary.

andrey-canon commented on edx2bigquery/s3_backend.py:
```diff
@@ -109,4 +113,67 @@ def get_tracking_log_objects(bucket_name, start_date):
+    # .zip Since edx-analytics-exporter uploads a .zip file containing all the generated data.
+    object_key_name = '{}/{}{}{}'.format(
+        getattr(edx2bigquery_config, 'SQL_DATA_BUCKET_PATH', ''),
+        getattr(edx2bigquery_config, 'SQL_FILE_NAME_PREFIX', ''),
```
constant

Squirrel18 replied:
This is a setting name, so, it's not neccesary.

andrey-canon commented on edx2bigquery/s3_backend.py:
```diff
@@ -109,4 +113,67 @@ def get_tracking_log_objects(bucket_name, start_date):
+        '.zip'
+    )
+    sql_data_local_dir = '{}/{}{}'.format(
+        getattr(edx2bigquery_config, 'SQL_LOCAL_FOLDER', ''),
```
constant

Squirrel18 replied:
This is a setting name, so, it's not neccesary.

andrey-canon commented on edx2bigquery/s3_backend.py:
```diff
@@ -109,4 +113,67 @@ def get_tracking_log_objects(bucket_name, start_date):
+    )
+    sql_data_local_dir = '{}/{}{}'.format(
+        getattr(edx2bigquery_config, 'SQL_LOCAL_FOLDER', ''),
+        getattr(edx2bigquery_config, 'SQL_FILE_NAME_PREFIX', ''),
```
constant

Squirrel18 replied:
This is a setting name, so, it's not neccesary.

andrey-canon commented on edx2bigquery/s3_backend.py:
```diff
@@ -109,4 +113,67 @@ def get_tracking_log_objects(bucket_name, start_date):
+    parent_path = os.path.abspath(os.path.join(my_path, os.pardir))
+    path_to_extract = '{}/{}'.format(
+        parent_path,
+        getattr(edx2bigquery_config, 'SQL_SOURCE_DATA_LOCAL_FOLDER', ''),
```
constant

Squirrel18 replied:
This is a setting name, so, it's not neccesary.

andrey-canon commented on edx2bigquery/s3_backend.py:
```diff
@@ -109,4 +113,67 @@ def get_tracking_log_objects(bucket_name, start_date):
+        getattr(edx2bigquery_config, 'SQL_SOURCE_DATA_LOCAL_FOLDER', ''),
+    )
+
+    if not os.path.exists(getattr(edx2bigquery_config, 'SQL_LOCAL_FOLDER', '')):
```
constant

Squirrel18 replied:
This is a setting name, so, it's not neccesary.

andrey-canon commented on edx2bigquery/s3_backend.py:
```diff
@@ -109,4 +113,67 @@ def get_tracking_log_objects(bucket_name, start_date):
+    )
+
+    if not os.path.exists(getattr(edx2bigquery_config, 'SQL_LOCAL_FOLDER', '')):
+        os.mkdir(getattr(edx2bigquery_config, 'SQL_LOCAL_FOLDER', ''))
```
constant

Squirrel18 replied:
This is a setting name, so, it's not neccesary.

Squirrel18 commented:
@andrey-canon Addressed.

Squirrel18 committed:
Adding support for local files for analyze_videos command.
Adding support for local files to person_course command.
Adding support for local files to axis2bq command.
Adding support for opaque keys.
Adding support for get SQL data from S3.
